### PR TITLE
Show more information when multiple `impl`s apply

### LINKED ISSUE #88284: Missing Type Annotations Error Lacks Suggestions
Given the following code: 

```rust
struct Foo {
    inner: u32,
}

struct Bar {
    inner: u32,
}

#[derive(Clone, Copy)]
struct Baz {
    inner: u32,
}

impl From<Baz> for Bar {
    fn from(other: Baz) -> Self {
        Self {
            inner: other.inner,
        }
    }
}

impl From<Baz> for Foo {
    fn from(other: Baz) -> Self {
        Self {
            inner: other.inner,
        }
    }
}

fn main() {
    let x: Baz = Baz { inner: 42 };

    // DOESN'T Compile: Multiple options!
    let y = x.into();

    let y_1: Foo = x.into();
    let y_2: Bar = x.into();

    let z_1 = Foo::from(y_1);
    let z_2 = Bar::from(y_2);

    // No type annotations needed, the compiler KNOWS the type must be `Foo`!
    let m = magic_foo(x);
}

fn magic_foo(arg: Baz) -> Foo {
    arg.into()
}

```

The current output is:

```
error[E0282]: type annotations needed
  --> src/main.rs:34:9
   |
34 |     let y = x.into();
   |         ^ consider giving `y` a type

error: aborting due to previous error

For more information about this error, try `rustc --explain E0282`.
```

Ideally the output should look like:

IMO the output could include:

* A reminder of the explicit type syntax (e.g. `let y: TYPE = x.into()`)
  * This is well suggested in `rustc --explain E0282`, but it might be helpful in the rustc output as well
* A suggestion of potential candidate (e.g. in this case: `Foo` or `Bar`)

### PR DESCRIPTION
**Pull request by estebank**
- When there are `impl`s without type params, show only those (to avoid showing overly generic `impl`s).
```
error[E0283]: type annotations needed
  --> $DIR/multiple-impl-apply.rs:34:9
   |
LL |     let y = x.into();
   |         ^     ---- type must be known at this point
   |
note: multiple `impl`s satisfying `_: From<Baz>` found
  --> $DIR/multiple-impl-apply.rs:14:1
   |
LL | impl From<Baz> for Bar {
   | ^^^^^^^^^^^^^^^^^^^^^^
...
LL | impl From<Baz> for Foo {
   | ^^^^^^^^^^^^^^^^^^^^^^
   = note: required for `Baz` to implement `Into<_>`
help: consider giving `y` an explicit type
   |
LL |     let y: /* Type */ = x.into();
   |          ++++++++++++
```

- Lower the importance of `T: Sized`, `T: WellFormed` and coercion errors, to prioritize more relevant errors. The pre-existing deduplication logic deals with hiding redundant errors better that way, and we show errors with more metadata that is useful to the user.

- Show `<SelfTy as Trait>::assoc_fn` suggestion in more cases.
```
error[E0790]: cannot call associated function on trait without specifying the corresponding `impl` type
  --> $DIR/cross-return-site-inference.rs:38:16
   |
LL |     return Err(From::from("foo"));
   |                ^^^^^^^^^^ cannot call associated function of trait
   |
help: use a fully-qualified path to a specific available implementation
   |
LL |     return Err(</* self type */ as From>::from("foo"));
   |                +++++++++++++++++++     +
```

Fix #88284.

### CODE REVIEW

File: tests/ui/inference/question-mark-type-infer.stderr [hunk 1]
```diff
@@ -1,14 +1,35 @@
...
+             where T: Eq, T: Hash, S: BuildHasher, S: Default;
+           - impl<T> FromIterator<T> for Arc<[T]>;
+           - impl<T> FromIterator<T> for BTreeSet<T>
+             where T: Ord;
+           - impl<T> FromIterator<T> for BinaryHeap<T>
+             where T: Ord;
+           - impl<T> FromIterator<T> for LinkedList<T>;
+           - impl<T> FromIterator<T> for Rc<[T]>;
+           - impl<T> FromIterator<T> for Vec<T>;
+           - impl<T> FromIterator<T> for VecDeque<T>;
```
Comment (compiler-errors) [inline tests/ui/inference/question-mark-type-infer.stderr:25]: I expect this to be quite overwhelming to a user. Why is it beneficial to show an exhaustive list of impls?

Comment (estebank) [reply tests/ui/inference/question-mark-type-infer.stderr:25]: My next step would be to special case the `From`, `Into`, `FromIterator` and `IntoIterator` impls to extract the right types for the structured suggestions and try to come up with some heuristic for which types you most likely intended. I can make the list shorter again and maybe display a handful and "and N more" like we do elsewhere.

Comment (estebank) [reply tests/ui/inference/question-mark-type-infer.stderr]: done

File: compiler/rustc_error_codes/src/error_codes/E0282.md [hunk 2]
```diff
@@ -3,7 +3,7 @@ The compiler could not infer a type and asked for a type annotation.
 Erroneous code example:
 
 ```compile_fail,E0282
-let x = "hello".chars().rev().collect();
+let x;
```
Comment (compiler-errors) [inline compiler/rustc_error_codes/src/error_codes/E0282.md:6]: Can you find a better representative for `E0282`? This is trivially ambiguous, but I think this error should have a more descriptive example.

Also, the description below doesn't match this example.

**Review by compiler-errors — COMMENTED** [review]
Can you fix up the history of this PR as well? Some of the intermediate changes like formatting and rebase fallout can probably be squashed down into the PRs that they're most closely related to.

**Comment by compiler-errors** [comment]
Needs rebase

**Review by wesleywiser — APPROVED** [review]
> tidy error: /checkout/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs: too many lines (3023) (add `// ignore-tidy-filelength` to the file to suppress this error)

r=me when CI is green 🙂

**Comment by estebank** [comment]
Split the file instead.

@bors r=wesleywiser

**Comment by estebank** [comment]
I'm not quite sure why there's a test output discrepancy here.

**Comment by estebank** [comment]
@bors r=wesleywiser

Figured what the problem was: There was a difference in the number of blanket impls that are present between WASM and Linux targets, the later having `gimli` impls in scope, which made the list fall in different sides of the "filter blanket impls" heuristic, which made the list ending up too long to display the relevant impls in the WASM target.

**Comment by estebank** [comment]
I feel like I'm chasing tests :(

@bors r=wesleywiser

**Comment by jamesmunns** [comment]
Thank you @estebank!

**Pull request merged**